iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/SnapshotDiffApplier.java

```diff
@@ -0,0 +1,468 @@
+
+      List<String> invalidDeleteDetails =
+          deletedIds.stream()
+              .filter(referencedIdsToRefs::containsKey)
+              .map(
+                  id ->
+                      String.format(
+                          "snapshot %s (referenced by: %s)",
+                          id, String.join(", ", referencedIdsToRefs.get(id))))
+              .collect(Collectors.toList());
+
+      if (!invalidDeleteDetails.isEmpty()) {
+        throw new InvalidIcebergSnapshotException(
+            String.format(
+                "Cannot delete snapshots that are still referenced by branches/tags: %s",
+                String.join("; ", invalidDeleteDetails)));
+      }
+    }
+
+    TableMetadata.Builder applyTo(TableMetadata metadata) {
+      TableMetadata.Builder builder = TableMetadata.buildFrom(metadata);
+
+      // Remove deleted snapshots
+      if (!deletedSnapshots.isEmpty()) {
+        Set<Long> deletedIds =
+            deletedSnapshots.stream().map(Snapshot::snapshotId).collect(Collectors.toSet());
+        builder.removeSnapshots(deletedIds);
+      }
+
+      // Remove stale branch references
+      metadata.refs().keySet().stream()
+          .filter(refName -> !providedRefs.containsKey(refName))
+          .forEach(builder::removeRef);
+
+      // Track existing snapshot IDs after deletions
+      Set<Long> existingAfterDeletion =
+          metadata.snapshots().stream().map(Snapshot::snapshotId).collect(Collectors.toSet());
+      deletedSnapshots.forEach(s -> existingAfterDeletion.remove(s.snapshotId()));
+
+      // Add unreferenced new snapshots
+      providedSnapshots.stream()
+          .filter(
+              s ->
+                  !existingAfterDeletion.contains(s.snapshotId())
+                      && !providedBranchIds.contains(s.snapshotId()))
+          .forEach(builder::addSnapshot);
+
+      // Set branch pointers
+      providedRefs.forEach(
+          (branchName, ref) -> {
+            Snapshot snapshot = providedById.get(ref.snapshotId());
+            if (snapshot == null) {
+              throw new InvalidIcebergSnapshotException(
+                  String.format(
+                      "Branch %s references non-existent snapshot %s",
+                      branchName, ref.snapshotId()));
+            }
+
+            if (existingAfterDeletion.contains(snapshot.snapshotId())) {
+              SnapshotRef existingRef = metadata.refs().get(branchName);
+              if (existingRef == null || existingRef.snapshotId() != ref.snapshotId()) {
+                builder.setRef(branchName, ref);
+              }
+            } else {
+              builder.setBranchSnapshot(snapshot, branchName);
+            }
+          });
+
+      return builder;
+    }
+
+    void recordMetrics(TableMetadata.Builder builder) {
+      int appendedCount =
+          (int)
+              regularSnapshots.stream()
+                  .filter(s -> !existingById.containsKey(s.snapshotId()))
+                  .count();
```

Comment:
Related to the above comment. We should have `regularSnapshots` as appended regular snapshots.